metadata-io/src/main/java/com/linkedin/metadata/entity/EntityUtils.java

```diff
@@ -278,4 +280,73 @@ public static List<SystemAspect> toSystemAspects(
 
     return systemAspects;
   }
+
+  /**
+   * Use the precalculated next version from system metadata if it exists, otherwise lookup the next
+   * version the normal way from the database
+   *
+   * @param aspectDao database access
+   * @param latestAspects aspect version 0 with system metadata
+   * @param urnAspects urn/aspects which we need next version information for
+   * @return map of the urn/aspect to the next aspect version
+   */
+  public static Map<String, Map<String, Long>> calculateNextVersions(
+      AspectDao aspectDao,
+      Map<String, Map<String, SystemAspect>> latestAspects,
+      Map<String, Set<String>> urnAspects) {
+    Map<String, Map<String, Long>> precalculatedVersions =
+        latestAspects.entrySet().stream()
+            .map(
+                entry ->
+                    Map.entry(
+                        entry.getKey(), convertSystemAspectToNextVersionMap(entry.getValue())))
+            .filter(entry -> !entry.getValue().isEmpty())
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    Map<String, Set<String>> missingAspectVersions =
+        urnAspects.entrySet().stream()
+            .flatMap(
+                entry ->
+                    entry.getValue().stream()
+                        .map(aspectName -> Pair.of(entry.getKey(), aspectName)))
+            .filter(
+                urnAspectName ->
+                    !precalculatedVersions
+                        .getOrDefault(urnAspectName.getKey(), Map.of())
+                        .containsKey(urnAspectName.getValue()))
+            .collect(
+                Collectors.groupingBy(
+                    Pair::getKey, Collectors.mapping(Pair::getValue, Collectors.toSet())));
+    Map<String, Map<String, Long>> databaseVersions =
+        missingAspectVersions.isEmpty()
+            ? Map.of()
+            : aspectDao.getNextVersions(missingAspectVersions);
+
+    // stitch back together the precalculated and database versions
+    return Stream.concat(
+            precalculatedVersions.entrySet().stream(), databaseVersions.entrySet().stream())
+        .collect(
+            Collectors.toMap(
+                Map.Entry::getKey,
+                Map.Entry::getValue,
+                (m1, m2) ->
+                    Stream.concat(m1.entrySet().stream(), m2.entrySet().stream())
+                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))));
+  }
+
+  /**
+   * Given a map of aspect name to system aspect, extract the next version if it exists
+   *
+   * @param aspectMap aspect name to system aspect map
+   * @return aspect name to next aspect version
+   */
+  private static Map<String, Long> convertSystemAspectToNextVersionMap(
+      Map<String, SystemAspect> aspectMap) {
+    return aspectMap.entrySet().stream()
+        .filter(entry -> entry.getValue().getVersion() == 0)
+        .map(entry -> Map.entry(entry.getKey(), entry.getValue().getSystemMetadataVersion()))
```

Comment:
for the aspect at version == 0, 
`entry.getValue().getSystemMetadataVersion()` will return the monotonically increasing version, but how is this the "next" version?
Isn't this the current version?

Reply:
My current understanding is
`version` is the custom logic version # which will be `== 0` for the latest aspect
`nextAspectVersion` is the monotonically increasing version which is exposed to clients as `version` in system metadata

Reply:
ok so the version in the system metadata is actually the next version that will be computed 👍